src/cliargs/core.lua

```diff
@@ -0,0 +1,623 @@
+        if not required[argument_cursor] then
+          if optargument.maxcount > 1 then
+            table.insert(values[optargument], curr_opt)
+          else
+            values[optargument] = curr_opt
+          end
+
+          if optargument.callback then
+            local status, err = optargument.callback(optargument.key, curr_opt)
+            if status == nil and err then
+              return on_error(err)
+            end
+          end
+        -- a regular argument:
+        else
+          local entry = required[argument_cursor]
+
+          values[entry] = curr_opt
+
+          if entry.callback then
+            local status, err = entry.callback(entry.key, curr_opt)
+
+            if status == nil and err then
+              return on_error(err)
+            end
+          end
+        end
+      end
+    end
+
+    local arg_count = argument_cursor
+
+    -- missing any required arguments, or too many?
+    if arg_count < #required or arg_count > #required + optargument.maxcount then
+      if optargument.maxcount > 0 then
+        return on_error(
+          "bad number of arguments: " ..
+          #required .. "-" .. #required + optargument.maxcount ..
+          " argument(s) must be specified, not " .. arg_count
+        )
+      else
+        return on_error(
+          "bad number of arguments: " ..
+          #required .. " argument(s) must be specified, not " .. arg_count
+        )
+      end
+    end
+
+    -- if necessary set the defaults for the last optional argument here
+    if optargument.maxcount > 1 and #values[optargument] == 0 then
+      values[optargument] = { optargument.default }
+    end
+
+    -- populate the results table
+    local results = {}
+
+    if optargument.key then
+      results[optargument.key] = values[optargument]
+    end
+
+    for _, entry in pairs(required) do
+      results[entry.key] = values[entry]
+    end
+
+    for _, entry in pairs(optional) do
+      if entry.key then
+        results[entry.key] = values[entry]
+      end
+
+      if entry.expanded_key then
+        results[entry.expanded_key] = values[entry]
+      end
+    end
+
+    if dump then
+      local msg = cli.printer.dump_internal_state(values)
+
+      cli.printer.print(msg)
+
+      return on_error("commandline dump created as requested per '--__DUMP__' option")
+    end
+
+    return results
+  end
+
+  --- Prints the USAGE message.
+  ---
+  --- @return {string}
+  ---         The USAGE message.
+  function cli:print_usage()
+    local msg = cli.printer.generate_usage()
+
+    if not silent then
+      cli.printer.print(msg)
+    end
+
+    return msg
+  end
+
+  --- Prints the HELP information.
+  ---
+  --- @return {string}
+  ---         The HELP message.
+  function cli:print_help()
+    local msg = ''
+
+    msg = msg .. cli.printer.generate_usage() .. '\n'
+    msg = msg .. cli.printer.generate_help()
+
+    if not silent then
+      cli.printer.print(msg)
+    end
+
+    return msg
+  end
```

Comment:
Someone may want to do
```lua
cli:set_silent(true)
...
cli:print_help() -- or cli:print_usage()
```
in which case, nothing will be printed out because `silent` is true.  This could be confusion since it's not doing what it says.  Maybe `print_usage` and `print_help` should not depend on `silent`.